handsontable/src/utils/parseTable.js

```diff
@@ -25,130 +27,551 @@ function isHTMLTable(element) {
+      return `<span style="mso-spacerun: yes">${'&nbsp;'.repeat(substring.length - 1)} </span>`;
+    })
+    .replace(/\t/gi, '&#9;');
+}
+
+/**
+ * Converts Handsontable's header coordinates into HTMLTableElement.tHead.
+ *
+ * @param {Core} instance The Handsontable instance.
+ * @param {object} config Configuration for building HTMLTableElement.tHead.
+ * @param {Array<number>} config.rows List of row indexes which should be taken into account when creating
+ * the HTMLTableElement.tHead.
+ * @param {Array<number>} config.columns List of column indexes which should be taken into account when creating
+ * the HTMLTableElement.tHead.
+ * @returns {Array<string>} List of HTMLElements stored as strings.
+ */
+function getHeadersHTMLByCoords(instance, config) {
+  const { rows, columns } = config;
+  const headers = rows.filter(rowIndex => rowIndex < 0);
+  const headersHTML = [];
+
+  if (headers.length === 0 || columns.length === 0) {
+    return [];
+  }
+
+  headers.forEach((rowIndex) => {
+    const tr = ['<tr>'];
+
+    for (let i = 0; i < columns.length; i += 1) {
+      const columnIndex = columns[i];
+      const headerCell = instance.getCell(rowIndex, columnIndex);
+      const colspan = headerCell?.getAttribute('colspan');
+      let colspanAttribute = '';
+
+      if (colspan) {
+        const parsedColspan = parseInt(colspan, 10);
+
+        colspanAttribute = ` colspan=${parsedColspan}`;
+        i += parsedColspan - 1;
+      }
+
+      tr.push(`<th${colspanAttribute}>${encodeHTMLEntities(instance.getColHeader(columnIndex, rowIndex))}</th>`);
+    }
+
+    tr.push('</tr>');
+    headersHTML.push(...tr);
+  });
+
+  return ['<thead>', ...headersHTML, '</thead>'];
+}
+
+/**
+ * Converts Handsontable's coordinates into list of values for cells being headers.
+ *
+ * @param {Core} instance The Handsontable instance.
+ * @param {object} config Configuration for building the cell value list.
+ * @param {Array<number>} config.rows List of row indexes which should be taken into account when creating the
+ * cell value list.
+ * @param {Array<number>} config.columns List of column indexes which should be taken into account when creating the
+ * cell value list.
+ * @returns {Array[]} List of displayed cell values.
+ */
+function getHeadersDataByCoords(instance, config) {
+  const headersData = [];
+  const { columns, rows } = config;
+  const headers = rows.filter(rowIndex => rowIndex < 0);
+
+  headers.forEach((rowIndex) => {
+    const tr = [];
+
+    for (let i = 0; i < columns.length; i += 1) {
+      const columnIndex = columns[i];
+      const headerCell = instance.getCell(rowIndex, columnIndex);
+      const colspan = headerCell?.getAttribute('colspan');
+
+      tr.push(instance.getColHeader(columnIndex, rowIndex));
+
+      if (colspan) {
+        const parsedColspan = parseInt(colspan, 10);
+
+        tr.push(...new Array(parsedColspan - 1).fill(''));
+        i += parsedColspan - 1;
+      }
     }
+
+    headersData.push(tr);
```

Comment:
```suggestion
    const rowData = [];

    for (let i = 0; i < columns.length; i += 1) {
      const columnIndex = columns[i];
      const headerCell = instance.getCell(rowIndex, columnIndex);
      const colspan = headerCell?.getAttribute('colspan');

      rowData.push(instance.getColHeader(columnIndex, rowIndex));

      if (colspan) {
        const parsedColspan = parseInt(colspan, 10);

        rowData.push(...new Array(parsedColspan - 1).fill(''));
        i += parsedColspan - 1;
      }
    }

    headersData.push(rowData);
```
I started playing with a code, and it started to be confused. Every time I saw a `tr` variable, I thought there would be TR elements, but there are collected column values for a single row.